diff -Naru uClibc-0.9.33.2_old/libc/stdio/_wcommit.c uClibc-0.9.33.2_new/libc/stdio/_wcommit.c
--- uClibc-0.9.33.2_old/libc/stdio/_wcommit.c	2012-05-15 15:20:09.000000000 +0800
+++ uClibc-0.9.33.2_new/libc/stdio/_wcommit.c	2016-02-23 15:05:45.662571544 +0800
@@ -20,7 +20,18 @@
 
 	__STDIO_STREAM_VALIDATE(stream);
 
-	if ((bufsize = __STDIO_STREAM_BUFFER_WUSED(stream)) != 0) {
+	/* Note: we do not write anything if write error has been detected.
+	 * Otherwise, stdio user has no way to prevent retries after
+	 * failed write - and some users do want to not have any retries!
+	 * IOW: if write errored out, neither fflush nor fclose should
+	 * try to write buffered data.
+	 * clearerr may be used to enable retry if needed.
+	 */
+
+	bufsize = __STDIO_STREAM_BUFFER_WUSED(stream);
+	if (bufsize != 0
+		&& !(stream->__modeflags & __FLAG_ERROR)
+		) {
 		stream->__bufpos = stream->__bufstart;
 		__stdio_WRITE(stream, stream->__bufstart, bufsize);
 	}
diff -Naru uClibc-0.9.33.2_old/libc/stdio/_WRITE.c uClibc-0.9.33.2_new/libc/stdio/_WRITE.c
--- uClibc-0.9.33.2_old/libc/stdio/_WRITE.c	2012-05-15 15:20:09.000000000 +0800
+++ uClibc-0.9.33.2_new/libc/stdio/_WRITE.c	2016-02-23 15:05:57.563866009 +0800
@@ -79,6 +79,7 @@
 			 */
 			if (errno != EINTR && errno != EAGAIN) {
 				/* do we have other "soft" errors? */
+				bufsize -= todo;
 				break;
 			}
 #ifdef __STDIO_BUFFERS
